### PR TITLE
Xtensa asm2 framework

### PR DESCRIPTION
**Pull request by andyross**
Second cut.  This one is ready for merge if anyone feels it's ready.   We can hold off though too: the integration bits are another day or two out, so this is just the primitives and unit tests

### CODE REVIEW

**Comment by andyross** [comment]
Forgot to note: I cleaned up the true positive checkpatch issues.  The remaining warnings are all due to assembly syntax.

**Comment by andyross** [comment]
Add first phase of integration plumbing.  hello_world now runs correctly.  Interrupts are next.

File: drivers/timer/xtensa_sys_timer.c
Comment (andrewboie) [inline drivers/timer/xtensa_sys_timer.c]: huzzah!

File: samples/xtensa_asm2/src/main.c
Comment (andrewboie) [inline samples/xtensa_asm2/src/main.c]: consider getting rid of this unused code block, or conditionally compiling with some debug Kconfig instead of #if 0.
If you really want it set up this way it's fine w/me, more of a suggestion.

File: samples/xtensa_asm2/src/main.c
Comment (andrewboie) [inline samples/xtensa_asm2/src/main.c]: same as above

File: samples/xtensa_asm2/src/main.c
Comment (andrewboie) [inline samples/xtensa_asm2/src/main.c]: please convert this to ztest so that success/failure is reported in the expected way and move this under tests/.
We want sanitycheck/test case framework to exercise this in qemu-xtensa or real hardware, since we appear to be able to determine success/failure at runtime.

**Review by tyrantlizard77 — COMMENTED** [review]
A lot of the ASM stuff is very obscure to me. I would be more comfortable about its correctness if we were exercising it in CI, which we currently aren't.

I am not super into the idea of having two parallel implementations of the asm domain for Xtensa. I think that since we are early in the 1.11 cycle we should just commit to the new one, so that it gets exercised in test case framework and sanitycheck.

I also must insist on the provided sample to be converted to a ztest case, or at least converted to an old-style test, such that it gets run by sanitycheck and reports its success/failure in a way that our test case infra understands.

**Comment by andyross** [comment]
Agreed that the test integration is needed.  Though this is going to need a stabilization phase as always.  So far it's been hugely useful to be able to run the primitives underneath the existing framework as unit tests, and that will probably continue to be true for a bit.  I'd like to keep it around at least through one release.

But sure: at the end of the process, and if we're OK with dropping the "dead" platforms in there (i.e. everything but qemu and ESP-32, our only supported board targets), then sure: I'd be happen to push a patch removing it all.

**Comment by andyross** [comment]
Next cut.  This is 100% complete now and passes all my own unit tests.  Runs a bunch of the broader tests (when setting the CONFIG_XTENSA_ASM2=y as a default), still fails on a few.

File: arch/xtensa/soc/esp32/include/_soc_inthandlers.h
Comment (lpereira) [inline arch/xtensa/soc/esp32/include/_soc_inthandlers.h]: I wonder why this is needed; can't something that uses `__builtin_ffs()` be used instead? This code must be in the ESP32 IRAM and compete with other things that must be there (compared with things that can live in flash and be used via the flash cache).

```C
int irq = _builtin_ffs(mask);
if (irq) {
        _sw_isr_table[irq].isr(_sw_isr_table[irq].arg);
       return 1<<irq;
}
return 0;
```

**Comment by andyross** [comment]
Regarding __builtin_ffs: the builtin is actually slower.  GCC emits code to call a function and roll the register windows instead of putting it inline.  There is a (technically optional, probably why GCC doesn't inline it) NSAU instruction (not quite exactly CLZ/FFS, but same general idea) that could be used.  But... if GCC would generate optimal code, this trick (binary searching the masks) would actually beat that too.  But it doesn't, because it won't emit a BBS instruction (test a static bit and branch in a single instruction) and instead wants to test a mask for the single bits, which defeats the gain.

So... yeah, some rework here for cycle optimization couldn't hurt.  Generating assembly for the selection would be an option.  Another trick would be to integrate it with the ISR table generation so you can have the static function pointer and argument inline in the code instead of loaded from memory.  That's another two cycles.

**Comment by andyross** [comment]
Oh, and I realized I didn't answer your question: that code is the basic idea, but you also have the problem that the INTERRUPT register will flag interrupts from all levels, not just the one your are servicing.  So you need to make sure that a level N interrupt checks and calls only functions from level N, or else you might accidentally call a low-priority ISR out of a high priority handler simply because they fired at the same time.  (Which is why I wanted to generate this code from core-isa.h in the first place).

**Comment by lpereira** [comment]
My main issue isn't much with interrupt latency -- but with code size.  At least on the ESP32, IRAM is quite limited (32KiB IIRC) and the whole kernel, drivers, and ISRs must be placed there. The reason being that that port currently runs entirely off IRAM (it's placed there by the 1st stage bootloader); and, when the flash cache is enabled, code that handles IRQs must be available. If some code must be paged in to the flash cache when an interrupt is being handled, a fault will occur.

This isn't bad (Zephyr is small enough that it fits comfortably in IRAM), but anything we can do to reduce the code size for something that's not optional is welcome.

**Comment by andyross** [comment]
Gotcha.  Yeah, a tighter loop wouldn't hurt.  To be fair: the last I checked this series as a whole knocked like 1100 bytes off the .text size, so there's a little room for waste.

**Comment by lpereira** [comment]
Yeah, 1KB of code is indeed a lot.  If anything, I'm fine with the way things are, given all the improvements made in this series.  It's just something that we can revisit in the future if space is at a premium.

File: arch/xtensa/Kconfig
Comment (andrewboie) [inline arch/xtensa/Kconfig]: I would like to see this default Y unless it breaks something

**Comment by andrewboie** [comment]
> dropping the "dead" platforms

Not sure I understand this.
Does the new code not work with the various xt-sim targets?

**Comment by andyross** [comment]
It only probably works with the xcc targets (at least assuming they are all windowed variants -- without register windows this definitely doesn't work).  Frankly I don't even have a way to enumerate the core-isa.h files for those to find out; they're s all buried behind the proprietary SDK.  As it stands it definitely supports only qemu_xtensa and ESP-32.

Again, the goal here was to have a reworked interrupt and swap framework for SMP, and it grew from there.  I'd love for this to eventually eat the world, but IMHO it's fine to have a "legacy" layer for the old/proprietary targets with vendor support and a more modern one for current work.

**Comment by andrewboie** [comment]
> I'd love for this to eventually eat the world, but IMHO it's fine to have a "legacy" layer for the old/proprietary targets with vendor support and a more modern one for current work.

Works for me

**Review by tyrantlizard77 — DISMISSED** [review]

File: arch/xtensa/include/xtensa-asm2-s.h
Comment (andyross) [inline arch/xtensa/include/xtensa-asm2-s.h]: This comment is wrong, "and set INTLEVEL" is actually not what it does (though it's what it *should* do, sigh).  Thus the "gotcha" comment and code above.  Will fix.

**Review by andyross — COMMENTED** [review]
Hm... reading through there's some stuff I have to fix before this merges.

And I'm looking at what is for sure a real bug with switching a brand new stack early (or something) that the Dining Philosophers test is triggering, and that needs to be fixed.

Hold off merging for now.  Will try to fix from New Zealand.

File: arch/xtensa/core/xtensa_intgen.py
Comment (andyross) [inline arch/xtensa/core/xtensa_intgen.py]: Another lying comment.  It takes the output of preprocessing the template file, not core-isa.h itself.

File: arch/xtensa/core/xtensa_intgen.py
Comment (lpereira) [inline arch/xtensa/core/xtensa_intgen.py]: Just a nit, but this script needs a copyright+license header.

File: arch/xtensa/core/xtensa_intgen.py
Comment (lpereira) [inline arch/xtensa/core/xtensa_intgen.py]: Python has a max() builtin that works pretty much with everything: you can pass various arguments or any iterable object. So this loop can be written as `max_level = max(ints_by_lvl)`.

File: arch/xtensa/core/xtensa_intgen.py
Comment (lpereira) [inline arch/xtensa/core/xtensa_intgen.py]: Not really necessary, but you can simplify this slightly by removing the previous `for lvl...` loop, and writing it like this: `emit_int_handler(sorted(ints_by_lvl.get(lvl, [])))`.  Maybe even remove the `max_level = max(...)` line as well.

Comment (andyross) [reply arch/xtensa/core/xtensa_intgen.py]: Yeah.  I write toddler python.  Guilty as charged.

File: samples/xtensa_asm2/CMakeLists.txt
Comment (nashif) [inline samples/xtensa_asm2/CMakeLists.txt]: move to samples/boards

**Review by nashif — CHANGES_REQUESTED** [review]

File: samples/xtensa_asm2/CMakeLists.txt
Comment (nashif) [reply samples/xtensa_asm2/CMakeLists.txt]: and this is more of a test than a sample, can we move this to tests/ please?

File: samples/xtensa_asm2/src/main.c
Comment (nashif) [reply samples/xtensa_asm2/src/main.c]: this code block should be removed.

File: samples/xtensa_asm2/src/main.c
Comment (nashif) [reply samples/xtensa_asm2/src/main.c]: agree

**Review by nashif — CHANGES_REQUESTED** [review]

**Comment by andyross** [comment]
Dust this off and push all the final fixes.  Finally got all of sanitycheck running on qemu_xtensa and a decent amount of breadth on manual esp32 testing.  Switching over to asm2 exclusively on those platforms, so the CI test will show a proper result using the new code.

I do see the comments on the sample/test code which is still the original.  Will address, though maybe it would be easier to remove that from the series if someone wants to merge immediately.

**Comment by nashif** [comment]
@rgundi can you please test this patch set with Intel S1000 platform?

**Comment by andyross** [comment]
Note that Xtensa SoCs other than ESP-32 and qemu (which is called "sample_controller" in our tree for reasons I don't understand) will still build with the legacy code.  That absolutely should be tested and shouldn't break.  But we'll want to migrate other supported platforms ASAP.   Requires running the one python script to generate the interrupt stubs against the per-SoC toolchain.  Bug me offline.

**Comment by andrewboie** [comment]
> which is called "sample_controller" in our tree for reasons I don't understand

The SOC is called this because the specific CPU build in the Cadence SDK it is targeting is also named "sample_controller".

**Comment by andyross** [comment]
And of course I check this morning and discover I pushed an unrebased tree.  Now it should work.

**Comment by nashif** [comment]
can we close this?

**Pull request closed without merging**
